### DISCUSSION TITLE
Dynamically load nested entities based on GraphQL query

### DISCUSSION THREAD

**Discussion by SteveMelons** [mikro-orm/mikro-orm#1264] [Q&A] [answered]
I am working with GraphQL and MikroORM. My GQL object types are essentially identical to my MirkoORM entities. Now I am trying to figure out how to dynamically initialize nested entities and collections depending on which fields were actually requested through GQL.

Example:
```typescript
@Entity()
export class User {
  @PrimaryKey()
  id!: string;

  @OneToMany(() => Post, (post) => post.creator)
  posts = new Collection<Post>(this);
}

@Entity()
export class Post{
  @PrimaryKey()
  id!: string;

  @ManyToOne()
  creator!: User;

  @OneToMany(() => Image, (image) => image.post)
  images = new Collection<Image>(this);
}

@Entity()
export class Image{
  @PrimaryKey()
  id!: string;

  @ManyToOne()
  post!: Post;
}
```

Now an incoming GQL query requesting a `User` could only fetch the `id` and everything would be fine. 
```graphql
query {
  user(id: "abc") {
    id
  }
}
```

But if the query requests the posts of that user, I need to initialize the `posts` collection from that user.
```graphql
query {
  user(id: "abc") {
    id
    posts {
      id
    }
  }
}
```

But if the query then requests the images of the posts aswell, I need to initialize the posts collection from the user and the images collection from each post.
```graphql
query {
  user(id: "abc") {
    id
    posts {
      id
      images {
        id
      }
    }
  }
}
```

I have a very ugly and not very safe hack for now, but it only works for initialzing one layer deep. It works by just taking the GQL request as a string and checking if it includes any of the strings passed into the function.

Example:
```typescript
const post = await em.findOne(Post,{ id },filterFields(info, ["creator"]));

const filterFields = (info: any, fields: string[]) => {
  // only keep fields which are contained in the GQL request
  return fields.filter((field) =>
    info.fieldNodes[0].selectionSet.loc.source.body.includes(field)
  );
};
```

Any ideas on how I could manage to pull this off in a somewhat simple and clean way?

I am using type-graphql and apollo in my backend. Maybe I could ask this question in the type-graphql/graphql/apollo discussion aswell but thought I would try here first.

**Comment by B4nan** [comment, marked as answer]
Can't help much with GQL, but have you seen this example project? 

https://github.com/driescroons/mikro-orm-graphql-example

Sounds like you are talking about the same functionality as the https://github.com/driescroons/graphql-fields-to-relations is for.

I guess we should have a section about GQL in the recipies section, but my experience with that is very limited...

**Reply by SteveMelons** [reply]
That is actually exactly what I was looking for. Thanks!
